// @glint/ember-tsc@1.0.9/lib/transform/template/map-template-contents.js
import { preprocess } from '@glimmer/syntax';
import { assert } from '../util.js';
import { codeFeatures } from './code-features.js';
import GlimmerASTMappingTree, { TemplateEmbedding, } from './glimmer-ast-mapping-tree.js';
/**
 * @glimmer/syntax parses identifiers as strings. Aside from meaning
 * we often have to reverse engineer location information for them
 * by hand, it also means we can't treat mappings from identifiers
 * consistently with how we treat mappings from other AST nodes.
 *
 * This class just gives us a uniform way to store identifiers
 * or other nodes as the `source` for a mapping.
 */
export class Identifier {
    constructor(name) {
        this.name = name;
        this.type = 'Identifier';
    }
}
/**
 * Given the text of a handlebars template (either standalone .hbs file, or the contents
 * of an embedded `<template>...</template>` within a .gts file), invokes the given callback
 * with a set of tools to emit mapped contents corresponding to
 * that template, tracking the text emitted in order to provide
 * a mapping of ranges in the input to ranges in the output.
 */
export function mapTemplateContents(template, { embeddingSyntax }, callback) {
    let ast = null;
    let errors = [];
    let lineOffsets = calculateLineOffsets(template, embeddingSyntax.prefix.length);
    try {
        ast = preprocess(template);
    }
    catch (error) {
        let message = getErrorMessage(error);
        let location;
        if (isHBSSyntaxError(error)) {
            location = {
                start: lineOffsets[error.hash.loc.first_line] + error.hash.loc.first_column,
                end: lineOffsets[error.hash.loc.last_line] + error.hash.loc.last_column,
            };
        }
        else {
            let match = /line (\d+) : column (\d+)/.exec(message);
            if (match) {
                let offset = lineOffsets[Number(match[1])] + Number(match[2]);
                location = { start: offset, end: offset };
            }
        }
        errors.push({ message, location });
    }
    let segmentsStack = [[]];
    let mappingsStack = [[]];
    let indent = '';
    let offset = 0;
    let directives = [];
    // Associates all content emitted during the given callback with the
    // given range in the template source and corresponding AST node.
    // If an exception is thrown while executing the callback, the error
    // will be captured and associated with the given range, and no content
    // will be emitted.
    let captureMapping = (hbsRange, source, allowEmpty, callback, codeFeaturesForNode) => {
        let start = offset;
        let mappings = [];
        let segments = [];
        segmentsStack.unshift(segments);
        mappingsStack.unshift(mappings);
        try {
            callback();
        }
        catch (error) {
            errors.push({ message: getErrorMessage(error), location: hbsRange });
            offset = start;
        }
        mappingsStack.shift();
        segmentsStack.shift();
        // If the offset didn't change (either because nothing was emitted
        // or because an exception was thrown), don't add a new node to the
        // mapping tree or flush any new content.
        if (start !== offset || allowEmpty) {
            let end = offset;
            let tsRange = { start, end };
            mappingsStack[0].push(new GlimmerASTMappingTree(tsRange, hbsRange, mappings, source, 
            // Prevent TS's semantic classifications (used for semantic highlighting, see
            // https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide) from being
            // source-mapped back to the glimmer template. These might be useful to reinstate at some
            // point in the future but by default tends to make the highlighting in gts files look wrong.
            codeFeaturesForNode ??
                augmentCodeFeaturesWithIgnoreDirectivesSupport(codeFeatures.withoutHighlight, hbsRange)));
            segmentsStack[0].push(...segments);
        }
    };
    /**
     * This function will conditionally augment the CodeInformation object
     * that's passed in for each mapping as a means to implement support for
     * `@glint-expect-error` directives.
     */
    function augmentCodeFeaturesWithIgnoreDirectivesSupport(features, hbsRange) {
        if (features.verification) {
            // If this code span requests verification (e.g. TS type-checking), then
            // we potentially need to decorate the `verification` value that we pass
            // back to Volar, in case we have active `glint-ignore/expect-error` directives
            // in active effect.
            let activeDirective;
            for (let directive of directives) {
                if (directive.areaOfEffect.start <= hbsRange.start &&
                    directive.areaOfEffect.end > hbsRange.start) {
                    if (!activeDirective) {
                        activeDirective = directive;
                    }
                    else {
                        // More than one directive applies here; glint-nocheck, if present, always wins.
                        if (directive.kind === 'nocheck') {
                            activeDirective = directive;
                        }
                    }
                }
            }
            if (!activeDirective) {
                return features;
            }
            if (activeDirective.kind === 'ignore' || activeDirective.kind === 'nocheck') {
                // We are currently in an ignored region of code, so don't
                // even bother performing any type-checking: override verification (i.e. type-checking) to false
                // for this mapping (note that the whole generated TS file will be type-checked but any
                // diagnostics in this region will be suppressed by Volar)
                return {
                    ...features,
                    verification: false,
                };
            }
            else if (activeDirective.kind === 'expect-error') {
                const expectErrorDirective = activeDirective;
                return {
                    ...features,
                    verification: {
                        shouldReport: () => {
                            // Keep track of any errors/diagnostics reported within this region of code...
                            expectErrorDirective.errorCount++;
                            // ...and suppress them from bubbling up as diagnostics/errrors/warnings.
                            return false;
                        },
                    },
                };
            }
        }
        return features;
    }
    let mapper = {
        indent() {
            indent += '  ';
        },
        dedent() {
            indent = indent.slice(2);
        },
        newline() {
            offset += 1;
            segmentsStack[0].push('\n');
        },
        text(value) {
            offset += value.length;
            segmentsStack[0].push(value);
        },
        synthetic(value) {
            if (value.length) {
                mapper.identifier(value, 0, 0);
            }
        },
        nothing(node, source = node) {
            captureMapping(mapper.rangeForNode(node), source, true, () => { });
        },
        identifier(value, hbsOffset, hbsLength = value.length) {
            let hbsRange = { start: hbsOffset, end: hbsOffset + hbsLength };
            let source = new Identifier(value);
            captureMapping(hbsRange, source, true, () => mapper.text(value));
        },
        forNode(node, callback, codeFeaturesForNode) {
            captureMapping(mapper.rangeForNode(node), node, false, callback, codeFeaturesForNode);
        },
        forNodeWithSpan(node, span, callback, codeFeaturesForNode) {
            captureMapping(mapper.rangeForNode(node, span), node, false, callback, codeFeaturesForNode);
        },
        error(message, location) {
            errors.push({ message, location });
        },
        directive(kind, commentNode, location, areaOfEffect) {
            const directive = {
                kind,
                commentNode,
                location,
                areaOfEffect,
                errorCount: 0,
            };
            directives.push(directive);
        },
        emitDirectivePlaceholders() {
            if (directives.length) {
                mapper.text('// begin directive placeholders');
                mapper.newline();
            }
            for (let directive of directives) {
                if (directive.kind !== 'expect-error') {
                    continue;
                }
                mapper.forNode(directive.commentNode, () => {
                    mapper.text(`// @ts-expect-error ${directive.kind}`);
                }, {
                    ...codeFeatures.withoutHighlight,
                    verification: {
                        shouldReport: () => {
                            // This determines whether we raise the placeholder "unused ts-expect-error" diagnostic.
                            // If errors were encountered in the region covered by the directive, then we suppress
                            // the "unused ts-expect-error" diagnostic here.
                            return directive.errorCount === 0;
                        },
                    },
                });
                mapper.newline();
                mapper.text(';');
                mapper.newline();
            }
            if (directives.length) {
                mapper.text('// end directive placeholders');
                mapper.newline();
            }
        },
        rangeForNode: buildRangeForNode(lineOffsets),
        rangeForLine: (line) => ({
            start: lineOffsets[line],
            end: lineOffsets[line + 1] ?? template.length,
        }),
    };
    callback(ast, mapper);
    assert(segmentsStack.length === 1);
    let code = segmentsStack[0].join('');
    let mapping = new GlimmerASTMappingTree({ start: 0, end: code.length }, {
        start: 0,
        end: embeddingSyntax.prefix.length + template.length + embeddingSyntax.suffix.length,
    }, mappingsStack[0], new TemplateEmbedding(), codeFeatures.all);
    return { errors, result: { code, directives, mapping } };
}
const LEADING_WHITESPACE = /^\s+/;
const TRAILING_WHITESPACE = /\s+$/;
function calculateLineOffsets(template, contentOffset) {
    let lines = template.split('\n');
    let total = contentOffset;
    let offsets = [contentOffset];
    for (let [index, line] of lines.entries()) {
        // lines from @glimmer/syntax are 1-indexed
        offsets[index + 1] = total;
        total += line.length + 1;
    }
    return offsets;
}
function buildRangeForNode(offsets) {
    return (node, span) => {
        let { loc } = node;
        if (span) {
            loc = span;
        }
        let start = offsets[loc.start.line] + loc.start.column;
        let end = offsets[loc.end.line] + loc.end.column;
        // This makes error reporting for illegal text nodes (e.g. alongside named blocks)
        // a bit nicer by only highlighting the content rather than all the surrounding
        // newlines and attendant whitespace
        if (node.type === 'TextNode') {
            let leading = LEADING_WHITESPACE.exec(node.chars)?.[0].length ?? 0;
            let trailing = TRAILING_WHITESPACE.exec(node.chars)?.[0].length ?? 0;
            if (leading !== node.chars.length) {
                start += leading;
                end -= trailing;
            }
        }
        return { start, end };
    };
}
function getErrorMessage(error) {
    return error?.message ?? '(unknown error)';
}
function isHBSSyntaxError(error) {
    if (typeof error === 'object' && !!error && 'hash' in error) {
        let { hash } = error;
        return typeof hash?.loc === 'object';
    }
    return false;
}
