// @handlebars/parser@2.2.2/dist/esm/parse.js
import parser from './parser.js';
import WhitespaceControl from './whitespace-control.js';
import * as Helpers from './helpers.js';
var baseHelpers = {};
for (var helper in Helpers) {
    if (Object.prototype.hasOwnProperty.call(Helpers, helper)) {
        baseHelpers[helper] = Helpers[helper];
    }
}
export function parseWithoutProcessing(input, options) {
    var _a, _b, _c;
    // Just return if an already-compiled AST was passed in.
    if (input.type === 'Program') {
        return input;
    }
    parser.yy = baseHelpers;
    // Altering the shared object here, but this is ok as parser is a sync operation
    parser.yy.locInfo = function (locInfo) {
        return new Helpers.SourceLocation(options && options.srcName, locInfo);
    };
    var squareSyntax;
    if (typeof ((_a = options === null || options === void 0 ? void 0 : options.syntax) === null || _a === void 0 ? void 0 : _a.square) === 'function') {
        squareSyntax = options.syntax.square;
    }
    else if (((_b = options === null || options === void 0 ? void 0 : options.syntax) === null || _b === void 0 ? void 0 : _b.square) === 'node') {
        squareSyntax = arrayLiteralNode;
    }
    else {
        squareSyntax = 'string';
    }
    var hashSyntax;
    if (typeof ((_c = options === null || options === void 0 ? void 0 : options.syntax) === null || _c === void 0 ? void 0 : _c.hash) === 'function') {
        hashSyntax = options.syntax.hash;
    }
    else {
        hashSyntax = hashLiteralNode;
    }
    parser.yy.syntax = {
        square: squareSyntax,
        hash: hashSyntax,
    };
    return parser.parse(input);
}
function arrayLiteralNode(array, loc) {
    return {
        type: 'ArrayLiteral',
        items: array,
        loc: loc,
    };
}
function hashLiteralNode(hash, loc) {
    return {
        type: 'HashLiteral',
        pairs: hash.pairs,
        loc: loc,
    };
}
export function parse(input, options) {
    var ast = parseWithoutProcessing(input, options);
    var strip = new WhitespaceControl(options);
    return strip.accept(ast);
}
