// @handlebars/parser@2.2.2/dist/cjs/parse.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseWithoutProcessing = parseWithoutProcessing;
exports.parse = parse;
var parser_js_1 = __importDefault(require("./parser.js"));
var whitespace_control_js_1 = __importDefault(require("./whitespace-control.js"));
var Helpers = __importStar(require("./helpers.js"));
var baseHelpers = {};
for (var helper in Helpers) {
    if (Object.prototype.hasOwnProperty.call(Helpers, helper)) {
        baseHelpers[helper] = Helpers[helper];
    }
}
function parseWithoutProcessing(input, options) {
    var _a, _b, _c;
    // Just return if an already-compiled AST was passed in.
    if (input.type === 'Program') {
        return input;
    }
    parser_js_1.default.yy = baseHelpers;
    // Altering the shared object here, but this is ok as parser is a sync operation
    parser_js_1.default.yy.locInfo = function (locInfo) {
        return new Helpers.SourceLocation(options && options.srcName, locInfo);
    };
    var squareSyntax;
    if (typeof ((_a = options === null || options === void 0 ? void 0 : options.syntax) === null || _a === void 0 ? void 0 : _a.square) === 'function') {
        squareSyntax = options.syntax.square;
    }
    else if (((_b = options === null || options === void 0 ? void 0 : options.syntax) === null || _b === void 0 ? void 0 : _b.square) === 'node') {
        squareSyntax = arrayLiteralNode;
    }
    else {
        squareSyntax = 'string';
    }
    var hashSyntax;
    if (typeof ((_c = options === null || options === void 0 ? void 0 : options.syntax) === null || _c === void 0 ? void 0 : _c.hash) === 'function') {
        hashSyntax = options.syntax.hash;
    }
    else {
        hashSyntax = hashLiteralNode;
    }
    parser_js_1.default.yy.syntax = {
        square: squareSyntax,
        hash: hashSyntax,
    };
    return parser_js_1.default.parse(input);
}
function arrayLiteralNode(array, loc) {
    return {
        type: 'ArrayLiteral',
        items: array,
        loc: loc,
    };
}
function hashLiteralNode(hash, loc) {
    return {
        type: 'HashLiteral',
        pairs: hash.pairs,
        loc: loc,
    };
}
function parse(input, options) {
    var ast = parseWithoutProcessing(input, options);
    var strip = new whitespace_control_js_1.default(options);
    return strip.accept(ast);
}
